Indicators/SharpeRatio.cs

```diff
@@ -0,0 +1,129 @@
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Calculation of the rolling moving average for the Sharpe Ratio (sr) developed by William F. Sharpe
+    /// You can optionally specify a different moving average type to be used in the computation.
+    ///
+    /// Reference: https://www.investopedia.com/articles/07/sharpe_ratio.asp
+    /// Formula: S(x) = (Rx - Rf) / stdDev(Rx)
+    /// Where:
+    /// S(x) - sharpe ratio of x
+    /// Rx - average rate of return for x
+    /// Rf - risk-free rate
+    /// </summary>
+    public class SharpeRatio : IndicatorBase<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// RollingWindow for storing historical values for percent change calculation
+        /// </summary>
+        private RollingWindow<IndicatorDataPoint> _values;
+
+        /// <summary>
+        /// StandardDeviation indicator for calculating the standard deviation over period
+        /// </summary>
+        private StandardDeviation _std;
+
+        /// <summary>
+        /// Stores the period of the historical percent changes used for calculating sharpe ratio
+        /// </summary>
+        public int SharpePeriod { get; }
+
+        /// <summary>
+        /// Stores the value of the risk-free rate of return for calculation
+        /// </summary>
+        public decimal RiskFreeRate { get; }
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
+
+        /// <summary>
+        /// Returns whether the indicator is properly initalized with data
+        /// </summary>
+        public override bool IsReady => _values.IsReady && _std.IsReady;
+
+        /// <summary>
+        /// Creates a new SharpeRatio indicator using the specified periods
+        /// </summary>
+		/// <param name="name">The name of this indicator</param>
+		/// <param name="sharpePeriod">Period of historical observation for sharpe ratio calculation</param>
+		/// <param name="riskFreeRate">Risk-free rate for sharpe ratio calculation</param>
+        public SharpeRatio(string name, int sharpePeriod, decimal riskFreeRate = 0.0m)
+            : base(name)
+        {
```

Comment:
You can use the indicator extensions to hold all the calculations here:
```csharp
_roc = new RateOfChange(1);
var std = new StandardDeviation(period).Of(_roc);
var sma = _roc.SMA(period);
_sharpeRatio = sma.Minus(riskFreeRate).Over(std);
```

and the body of `ComputeNextValue` will be reduced to:
```csharp
_roc.Update(input);
return _sharpeRatio;
```